File: lib-android/src/main/java/com/pranav/lib_android/util/FileUtil.java
package com.pranav.lib_android.util;

import android.content.Context;

import java.nio.charset.Charset;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;

import org.apache.commons.io.FileUtils;

public class FileUtil {

    private static Context mContext;

    public static void initializeContext(Context context) {
        mContext = context;
    }

    public static boolean createDirectory(String path) {
        return new File(path).mkdir();
    }
    
    public static void writeFile(InputStream in, String path) {
        File file = new File(path);
        FileUtils.copyInputStreamToFile(in, file);
    }

    public static void writeFile(String path, String content) throws IOException {
        File file = new File(path);
        file.getParentFile().mkdirs();
        if (!file.exists()) file.createNewFile();
        FileUtils.write(file, content, Charset.defaultCharset());
    }

    public static String readFile(File file) throws IOException {
        return FileUtils.readFileToString(file, Charset.defaultCharset());
    }

    public static void deleteFile(String path) {
        File file = new File(path);

        if (!file.exists()) return;

        if (file.isFile()) {
            file.delete();
            return;
        }

        File[] fileArr = file.listFiles();

        if (fileArr != null) {
            for (File subFile : fileArr) {
                if (subFile.isDirectory()) {
                    deleteFile(subFile.getAbsolutePath());
                }

                if (subFile.isFile()) {
                    subFile.delete();
                }
            }
        }

        file.delete();
    }

    public static String getFileName(String path) {
        String[] splited = path.split("/");
        return splited[splited.length - 1];
    }

    private static String getDataDir() {
        return mContext.getExternalFilesDir(null).getAbsolutePath();
    }

    public static String getJavaDir() {
        return getDataDir() + "/java/";
    }

    public static String getBinDir() {
        return getDataDir() + "/bin/";
    }
    
    public static String getCacheDir() {
        // write caches to external storage because we don't want android system to delete index files
        return getDataDir() + "/cache/";
    }

    public static String getClasspathDir() {
        return getDataDir() + "/classpath/";
    }
}

File: lib-android/src/main/java/com/pranav/lib_android/util/FileUtil.kt
package com.pranav.lib_android.util

import android.content.Context

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException

class FileUtil {
    companion object {
    private lateinit var mContext: Context

    @JvmStatic
    fun initializeContext(context: Context) {
        mContext = context
    }

    @JvmStatic
    fun createDirectory(path: String) = File(path).mkdir()

    @JvmStatic
    fun writeFile(path: String, content: String) {
        val file = File(path)
        file.getParentFile().mkdirs()
        if (!file.exists()) file.createNewFile()
        file.writeText(content)
    }

    @JvmStatic
    fun writeBytes(path: String, bytes: ByteArray) {
        File(path).writeBytes(bytes)
    }

    @JvmStatic
    fun readFile(file: File) = file.readText()

    @JvmStatic
    fun asByteArray(in: InputStream) = in.readBytes()

    @JvmStatic
    fun deleteFile(path: String) {
        File(path).deleteRecursively()
    }

    @JvmStatic
    fun getFileName(path: String): String {
        val splited = path.split("/")
        return splited[splited.length - 1]
    }

    @JvmStatic
    private fun getDataDir() = mContext.getFilesDir()

    @JvmStatic
    fun getJavaDir() = getDataDir() + "/java/"

    @JvmStatic
    fun getBinDir() = getDataDir() + "/bin/"

    @JvmStatic
    fun getCacheDir() = getDataDir() + "/cache/"

    @JvmStatic
    fun getClasspathDir() = getDataDir() + "/classpath/"
}
